### PR TITLE
DAOS-15591 client: allow missing mode in open with O_CREAT

### PR DESCRIPTION
**Pull request by wiliamhuang**
Normally mode is required in open()/openat() when O_CREAT in open flag. glibc allows missing mode in open with O_CREAT although permission could be undefined. Mode in glibc open() normally carries permission information. The mode in dfs_open() carries both permission and type. We need to clear type bits and set desired type before passing mode to dfs_open(). openat() calls open_common() in libpil4dfs. A test for open() is added. "__open" is called to work around compiling flag "-D_FORTIFY_SOURCE=3".

Features: pil4dfs

Required-githooks: true

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate owners.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

**Review by mchaarawi — CHANGES_REQUESTED** [review]
any PR that fixes a bug should add a test.

**Comment by wiliamhuang** [comment]
> any PR that fixes a bug should add a test.

I intended to add a test in dfuse_test.c. It causes compiling error due to the correctness check with compiling flag "-D_FORTIFY_SOURCE=3". I also requested Cedric to test this PR.
If you are fine with it, I can add the test in  dfuse_test.c and remove compiling flag "-D_FORTIFY_SOURCE=3".

**Comment by wiliamhuang** [comment]
> > any PR that fixes a bug should add a test.
> 
> I intended to add a test in dfuse_test.c. It causes compiling error due to the correctness check with compiling flag "-D_FORTIFY_SOURCE=3". I also requested Cedric to test this PR. If you are fine with it, I can add the test in dfuse_test.c and remove compiling flag "-D_FORTIFY_SOURCE=3".

With  "-D_FORTIFY_SOURCE=3", mode must be provided when O_CREAT is used in open(). This is checked by the compiler.

**Comment by knard-intel** [comment]
> > any PR that fixes a bug should add a test.
> 
> I intended to add a test in dfuse_test.c. It causes compiling error due to the correctness check with compiling flag "-D_FORTIFY_SOURCE=3". I also requested Cedric to test this PR. If you are fine with it, I can add the test in dfuse_test.c and remove compiling flag "-D_FORTIFY_SOURCE=3".

Do not think that removing this compiling flag for dfuse_test.c should be done.
Another way, could be to create a dedicated test as I have done for [DAOS-14348](https://github.com/daos-stack/daos/pull/13940/files#diff-f8498922c5c475ae358d753fd85d0c6be4b48e38753e4ad1458960e5b866a142) and remove this flag only for this test.

File: src/client/dfuse/pil4dfs/int_dfs.c [hunk 1]
```diff
@@ -1979,8 +1979,9 @@ open_common(int (*real_open)(const char *pathname, int oflags, ...), const char
 	}
 	/* file/dir should be handled by DFS */
 	if (oflags & O_CREAT) {
-		rc = dfs_open(dfs_mt->dfs, parent, item_name, mode | S_IFREG, oflags & (~O_APPEND),
-			      0, 0, NULL, &dfs_obj);
+		/* clear the bits for types first */
+		rc = dfs_open(dfs_mt->dfs, parent, item_name, (mode & (~S_IFMT)) | S_IFREG,
+			      oflags & (~O_APPEND), 0, 0, NULL, &dfs_obj);
```
Comment (mchaarawi) [inline src/client/dfuse/pil4dfs/int_dfs.c:1984]: I am confused by this in the first place.. what if we are creating a directory? i.e. why do we have | S_IFREG ?

File: src/client/dfuse/pil4dfs/int_dfs.c [hunk 2]
```diff
@@ -1979,8 +1979,9 @@ open_common(int (*real_open)(const char *pathname, int oflags, ...), const char
 	}
 	/* file/dir should be handled by DFS */
 	if (oflags & O_CREAT) {
-		rc = dfs_open(dfs_mt->dfs, parent, item_name, mode | S_IFREG, oflags & (~O_APPEND),
-			      0, 0, NULL, &dfs_obj);
+		/* clear the bits for types first */
+		rc = dfs_open(dfs_mt->dfs, parent, item_name, (mode & (~S_IFMT)) | S_IFREG,
```
Comment (mchaarawi) [inline src/client/dfuse/pil4dfs/int_dfs.c:1983]: another question, what if someone really specifies an invalid mode_t and we need to return an error?

Comment (wiliamhuang) [reply src/client/dfuse/pil4dfs/int_dfs.c:1983]: mode is not checked here. dfs_open needs to check the mode.

File: src/client/dfuse/pil4dfs/int_dfs.c (hunk 1 above)
Comment (wiliamhuang) [reply src/client/dfuse/pil4dfs/int_dfs.c:1984]: I thought mkdir is normal used for creating dir. I assumed open is not used to create a directory.
I will add the code to handle "O_DIRECTORY".

Comment (mchaarawi) [reply src/client/dfuse/pil4dfs/int_dfs.c:1984]: im not familiar of when we end up here.. im just referring to your comment on line 1980

File: src/client/dfuse/pil4dfs/int_dfs.c (hunk 2 above)
Comment (wiliamhuang) [reply src/client/dfuse/pil4dfs/int_dfs.c:1983]: There is no type bit in mode in glibc open(). Only permission bit there.
dfs_open() combines permission and type bits.

File: src/client/dfuse/pil4dfs/int_dfs.c (hunk 1 above)
Comment (mchaarawi) [reply src/client/dfuse/pil4dfs/int_dfs.c:1984]: to be clear if this is just for files, that's fine, but your comment is wrong.
I also assumed open_common() is for both files and dirs, but his is your code, so you should clarify

Comment (wiliamhuang) [reply src/client/dfuse/pil4dfs/int_dfs.c:1984]: I will run some tests with O_DIRCTORY and O_PATH combined with O_CREAT to check the behavior of glibc on regular file system. Then decide how to proceed about this issue.

File: src/client/dfuse/pil4dfs/int_dfs.c [hunk 3]
```diff
@@ -1947,6 +1947,11 @@ open_common(int (*real_open)(const char *pathname, int oflags, ...), const char
 
 	if (!is_target_path)
 		goto org_func;
+	if (oflags & O_CREAT && (oflags & O_DIRECTORY || oflags & O_PATH)) {
+		/* Create a dir is not supported. */
+		errno = ENOENT;
+		return (-1);
+	}
```
Comment (mchaarawi) [inline src/client/dfuse/pil4dfs/int_dfs.c:1954]: is there any chance we can be in this function if creating a directory? i.e. should this be an assertion that this is not a dir? also what about symlinks?

**Review by mchaarawi — COMMENTED** [review]
I think we should land this after discussion in the WG. since dfs_open() supports files, dirs, symlinks, we should at least just set that part of the mode bits since we know that information. the other bits will be undefined as it is now.

**Comment by wiliamhuang** [comment]
> I think we should land this after discussion in the WG. since dfs_open() supports files, dirs, symlinks, we should at least just set that part of the mode bits since we know that information. the other bits will be undefined as it is now.

In my impression, mkdir()/mkdirat() should be used to create a dir. symlink() should be called for creating a symbolic link. I am not sure we want to support creating directory with open()/openat().
It seems creating a directory with open() works on some filesystesms but does not work on others.

File: src/client/dfuse/pil4dfs/int_dfs.c (hunk 3 above)
Comment (wiliamhuang) [reply src/client/dfuse/pil4dfs/int_dfs.c:1954]: Normally mkdir()/mkdirat() should be used to create a dir. However, we have no control over users' applications.
/scratch on boro (nfs4) returns -1 (errno = ENOENT), so I did the same here.
I am not aware that there is a flag for creating a symlink in open().

Comment (wiliamhuang) [reply src/client/dfuse/pil4dfs/int_dfs.c:1954]: @knard-intel Thank you for your information!
This document is in the section of "BUGS". We are not necessary to follow this behavior.
Shall we actually create a directory or return -1 (errno = ENOENT) for "O_DIRECTORY"?

Comment (knard-intel) [reply src/client/dfuse/pil4dfs/int_dfs.c:1954]: > @knard-intel Thank you for your information! This document is in the section of "BUGS". We are not necessary to follow this behavior. Shall we actually create a directory or return -1 (errno = ENOENT) for "O_DIRECTORY"?

You are right, from my side returning -1 (errno = ENOENT) seems the best choice.

**Review by knard-intel — APPROVED** [review]

**Review by johannlombardi — APPROVED** [review]

**Pull request merged**